lib/spack/spack/spec.py

```diff
@@ -1440,8 +1440,13 @@ def full_hash(self, length=None):
                 self.to_node_dict(hash_function=lambda s: s.full_hash()),
                 default_flow_style=True, width=maxint)
             package_hash = self.package.content_hash()
-            sha = hashlib.sha256(yaml_text.encode('utf-8') + package_hash)
-            self._full_hash = base64.b32encode(sha.digest()).lower()
+            sha = hashlib.sha1(yaml_text.encode('utf-8') + package_hash)
+
+            b32_hash = base64.b32encode(sha.digest()).lower()
+            if sys.version_info[0] >= 3:
+                b32_hash = b32_hash.decode('utf-8')
+
+            self._full_hash = b32_hash
```

Comment:
This should only cache if the spec is concrete, like the `dag_hash` does.  I suspect that in practice, it always will be, but why not make it mirror dag_hash?

Reply:
Well, this function raises an exception immediately if the spec is not concrete, so your suspicion is correct :smile: and I think we could be good here.

Reply:
Good point!  Duh!

Reply:
Oh let's not start pointing out all our "duh" moments or I'll start feeling bad.

Thanks for the review and merge!